python/tempo/intervals.py

```diff
@@ -0,0 +1,626 @@
+from __future__ import annotations
+
+from typing import Optional
+
+from pyspark.sql.dataframe import DataFrame
+import pyspark.sql.functions as f
+from pyspark.sql.window import Window
+
+
+class IntervalsDF:
+    """
+    This object is the main wrapper over a `Spark DataFrame`_ which allows a
+    user to parallelize computations over snapshots of metrics for intervals
+    of time defined by a start and end timestamp and various dimensions.
+
+    The required dimensions are `series` (list of columns by which to
+    summarize), `metrics` (list of columns to analyze), `start_ts` (timestamp
+    column), and `end_ts` (timestamp column). `start_ts` and `end_ts` can be
+    epoch or TimestampType.
+
+    .. _`Spark DataFrame`: https://spark.apache.org/docs/latest/api/python/reference/pyspark.sql/api/pyspark.sql.DataFrame.html
+
+    """
+
+    def __init__(
+        self,
+        df: DataFrame,
+        start_ts: str,
+        end_ts: str,
+        series: list[str],
+        metrics: list[str],
```

Comment:
@R7L208 - I'm not sure that we need to explicitly define the metrics. These can be inferred from the underlying DF. This is something I've been doing for the v0.2 refactor is to have explicit the structural columns of the TSDF, and then implicitly identify the metrics based on that.

Reply:
For example you can define a `metric_cols` property that effectively returns the names of numeric columns that are not part of the structural set (`start_ts`, `end_ts`, `series`)

Reply:
I pushed an update to implement these changes.